packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -0,0 +1,369 @@
+  }
+}
+
+// Used for accounting for which annotation is active inside of the MouseTracker.
+class _TrackedAnnotation {
+  _TrackedAnnotation(this.annotation);
+
+  final MouseDetectorAnnotation annotation;
+
+  /// Tracks devices that are currently active for this annotation.
+  ///
+  /// If the mouse pointer corresponding to the integer device ID is
+  /// present in the Set, then it is currently inside of the annotated layer.
+  ///
+  /// This is used to detect layers that used to have the mouse pointer inside
+  /// them, but now no longer do (to facilitate exit notification).
+  Set<int> activeDevices = Set<int>();
+}
+
+/// Describes a function that finds an annotation given an offset in logical
+/// coordinates.
+///
+/// It is used by the [MouseTracker] to fetch annotations for the mouse
+/// position.
+typedef MouseDetectorAnnotationFinder = MouseDetectorAnnotation Function(Offset offset);
+
+/// Keeps state about which objects are interested in tracking mouse positions
+/// and notifies them when a mouse pointer enters, moves, or leaves an annotated
+/// region that they are interested in.
+///
+/// Owned by the [RendererBinding] class.
+class MouseTracker {
+  /// Creates a mouse tracker to keep track of mouse locations.
+  ///
+  /// All of the parameters must not be null.
+  MouseTracker(PointerRouter router, this.annotationFinder)
+      : assert(router != null),
+        assert(annotationFinder != null) {
+    router.addGlobalRoute(_handleEvent);
+  }
+
+  /// Used to find annotations at a given logical coordinate.
+  final MouseDetectorAnnotationFinder annotationFinder;
+
+  // The collection of annotations that are currently being tracked. They may or
+  // may not be active, depending on the value of _TrackedAnnotation.active.
+  final Map<MouseDetectorAnnotation, _TrackedAnnotation> _trackedAnnotations = <MouseDetectorAnnotation, _TrackedAnnotation>{};
+
+  /// Track an annotation so that if the mouse enters it, we send it events.
+  ///
+  /// This is typically called when the [AnnotatedRegion] containing this
+  /// annotation has been added to the layer tree.
+  void attachAnnotation(MouseDetectorAnnotation annotation) {
+    _trackedAnnotations[annotation] = _TrackedAnnotation(annotation);
+    // Schedule a check so that we test this new annotation to see if the mouse
+    // is currently inside its region.
+    _scheduleMousePositionCheck();
+  }
+
+  /// Stops tracking an annotation, indicating that it has been removed from the
+  /// layer tree.
+  ///
+  /// If the associated layer is not removed, and receives a hit, then
+  /// [collectMousePositions] will assert the next time it is called.
+  void detachAnnotation(MouseDetectorAnnotation annotation) {
+    final _TrackedAnnotation trackedAnnotation = _findAnnotation(annotation);
+    assert(trackedAnnotation != null, "Tried to detach an annotation that wasn't attached: $annotation");
+    if (trackedAnnotation.activeDevices.isNotEmpty) {
```

Comment:
The if seems unnecessary. Wouldn't the for loop just do nothing if the list is empty?